Informed RRT*

gammell opened the pull request:
Ok,
So here is the changes to use informed sampling in RRT\* (i.e., Informed RRT\*). While I was at it, I cleaned a thing or two up in the RRT\* code, so I'll be going through to put some inline comments about those as well.

I'm working to "relax" informed sampling to handle multiple starts/goals (i.e., GOAL_STATE**S**), so I'll submit that after this PR, hopefully before the next release.

gammell commented on src/ompl/geometric/planners/rrt/RRTstar.h:
```diff
@@ -123,6 +123,18 @@ namespace ompl
+            {
+                return rewireFactor_;
+            }
+
```
Since the constants in RRT* are of the form A > B, where there's (only) a closed form for B, I added the ability to set the multiplicative factor to define A in terms of B.
Inspired by what the FMT* code does.

gammell commented on src/ompl/geometric/planners/rrt/RRTstar.h:
```diff
@@ -148,7 +160,12 @@ namespace ompl
+                constrained to pass the the \e vertex  is greater than the current solution.
+                As the ancestory of the descendent vertices can change at anytime, this means that the removed
+                descendents may actually be capable of later providing a better solution once
+                their incoming path passes through a different vertex (e.g., a change in homotopy class) */
```
Expanded documentation on the limitations of the current pruning method.

I don't mean to say that there's anything wrong with this, I just wanted to make it clear where it can be overly aggressive. I guess I could make a safer pruning function (i.e., it only removes a vertex if all its descendents are to be removed too), but I don't know if that should replace this function or live in parallel.

gammell commented on src/ompl/geometric/planners/rrt/RRTstar.h:
```diff
@@ -173,17 +190,57 @@ namespace ompl
+            {
+                return useKNearest_;
+            }
+
```
I expanded RRT* to have both an r-disc and k-nearest variant.

mamoll replied:
Maybe this can be done by using the code in ompl/src/ompl/geometric/planners/prm/ConnectionStrategy.h?

gammell replied:
To my untrained eye, `prm/ConnectionStrategy.h` appears fairly tailored to PRM? There is also the problem that for the r-disc variants, the RRT\* bound is different than the PRM\* bound...

mamoll replied:
Ok, just leave as-is. I thought there might have been an opportunity to unify some code.

gammell commented on src/ompl/geometric/planners/rrt/RRTstar.h:
```diff
@@ -173,17 +190,57 @@ namespace ompl
             {
-                return boost::lexical_cast<std::string>(bestCost_);
+                return bestCost_;
             }
```
I moved the planner progress properties to private functions, as per Ioan's request, but I did make the raw data available via `numIterations()` and `getBestCost()`, as I know I have some needs for it...

isucan replied:
This sounds good, but do you need this as a PlannerProgressProperty? Properties must return string, so maybe duplicate the function? Add a suffix "Property" to the function name, leave the private version as returning string and return the actual cost / nr iterations in public functions?

gammell replied:
Yup, already done, though I'll clean up the names as requested.

gammell commented on src/ompl/geometric/planners/rrt/RRTstar.h:
```diff
@@ -219,6 +276,12 @@ namespace ompl
+
+            /** \brief Generate a sample */
+            void sampleUniform(base::State *statePtr);
+
```
Helper functions to simplify the creation/use of different samplers.

gammell commented on src/ompl/geometric/planners/rrt/RRTstar.h:
```diff
@@ -243,6 +306,9 @@ namespace ompl
 
+            /** \brief Gets the neighbours of a given motion, using either k-nearest of radius as appropriate. */
+            void getNeighbors(Motion *motion, std::vector<Motion*> &nbh);
+
```
Helper function to simplify the choice of r-disc/k-nearest variants.

isucan replied:
can this be const?

gammell replied:
Good eye. :smile:

gammell commented on src/ompl/geometric/planners/rrt/RRTstar.h:
```diff
@@ -256,14 +322,26 @@ namespace ompl
+            /** \brief Computes the solution cost heuristically as the cost to come from start to the motion plus
+                 the cost to go from the motion to the goal. If \e estimate is true, a heuristic estimate of the
+                 cost to come is used; otherwise, the current cost to come is used. */
+            base::Cost solutionHeuristic(const Motion *motion, const bool estimate = true) const;
```
The previous `costToGo` function was poorly named. It calculated more than cost-to-go, but in fact an estimate of an entire solution path constrained to go through that vertex by adding a cost-to-go estimate to the cost-to-come (current or estimated). Renamed as `solutionHeuristic` and cleaned up the documentation.

gammell commented on src/ompl/geometric/planners/rrt/RRTstar.h:
```diff
@@ -256,14 +322,26 @@ namespace ompl
+
+            /** \brief Computes the cost-to-go heuristically for goal states and goal regions using the motionCost given by the optimization objective.*/
+            base::Cost defaultCostToGoHeuristic(const base::State *state, const base::Goal *goal) const;
+
```
Separated the cost-to-go component from the `solutionHeuristic` function.

gammell commented on src/ompl/geometric/planners/rrt/RRTstar.h:
```diff
@@ -256,14 +322,26 @@ namespace ompl
+            This is a sort of virtual pruning that avoids the problem of removing promising descendents simply
+            because of the quality of their ancestors.
+            \todo Replace with an actually pruning method. */
+            void countNumberOfVerticesWorseThanSoln();
```
If I were to rewrite the pruning function as described above, this function could be removed. It's like a sort of *virtual*-prune.

gammell commented on src/ompl/geometric/planners/rrt/src/RRTstar.cpp:
```diff
@@ -162,12 +205,6 @@ ompl::base::PlannerStatus ompl::geometric::RRTstar::solve(const base::PlannerTer
-    base::Cost bestCost    = opt_->infiniteCost();
-
-    bestCost_ = opt_->infiniteCost();
-
```
OK.
I would really like someone to double check me here, as I have tried to clean up this weird `bestCost`/`bestCost_` thing. Here both variables were set to the same thing.

jvgomez replied:
Last summer I was trying to remove this as well but at the end I didn't. In my case I was using both for the CForest framework, but now I agree that bestCost can be simply removed. It would be nice if you can run the CForest demo/tests with these changes to be sure that nothing is broken.

gammell commented on src/ompl/geometric/planners/rrt/src/RRTstar.cpp:
```diff
@@ -413,20 +458,23 @@ ompl::base::PlannerStatus ompl::geometric::RRTstar::solve(const base::PlannerTer
+                    if (opt_->isCostBetterThan(goalMotions_[i]->cost, bestCost_))
                     {
-                        bestCost = goalMotions_[i]->cost;
-                        bestCost_ = bestCost;
```
Here both variables are also set to the same thing. These are the only two places that either variable are assigned, so therefore, we can just remove `bestCost`.

isucan replied:
That seems to be the case, yes. +1

gammell commented on src/ompl/geometric/planners/rrt/src/RRTstar.cpp:
```diff
@@ -632,14 +709,112 @@ void ompl::geometric::RRTstar::deleteBranch(Motion *motion)
+    }
+    else if (pdef_->getGoal()->hasType(base::GOAL_REGION) == true)
+    {
+        return base::goalRegionCostToGo(state, goal);
```
My understanding from the documentation is that this function only works for `GOAL_REGION` types? Hence the switch. I guess I could handle the GOAL_STATE**S** case by finding the min over the countable set. I will update that when I push changes back from your comments. :smile:

gammell replied:
So, I see now that `GoalState` and `GoalStates` are both derived from `GoalSampleableRegion`, which is itself derived from `GoalRegion`...

I think I need some guidance.

Does that mean `base::goalRegionCostToGo` **is** valid for goals of type `GoalState`? Does it just do the same as what I was doing with the switch? In the case of `GoalStates`, does it find the min over the states or does it assume that the cost-to-go calculated with any one is the same?... I will have to look into it more, but this could really clean up some of my other code. :smile:

gammell replied:
Got this part sorted. Updates to the PR to follow shortly.

gammell commented on src/ompl/geometric/planners/cforest/src/CForest.cpp:
```diff
@@ -75,8 +75,11 @@ void ompl::geometric::CForest::addPlannerInstanceInternal(const base::PlannerPtr
+            planner->params()["focus_search"] = prune_;
+        else
+            OMPL_WARN("%s does not appear to support search focusing.", planner->getName().c_str());
+
```
More about this change explained in the RRTstar code, but my general thought here is to turn on a generic parameter called *focus_search*, instead of something as specifically named as *prune*.

This should make it conceptually cleaner for other planners to support the CFOREST framework even if they focus their search in ways other than pruning the tree.

gammell replied:
Maybe the parameter should even be called something like *enable_cforest_support* or something?

isucan replied:
I think I like focus_search. We just need to make sure we use the name consistently. Please add a comment explaining what the param means.

gammell commented on src/ompl/geometric/planners/rrt/RRTstar.h:
```diff
@@ -148,42 +160,125 @@ namespace ompl
+            /** \brief Set the fractional change in solution cost necessary for pruning to occur, i.e.,
+                prune if the new solution is at least X% better than the old solution.
+                (e.g., 0.0 will prune after every new solution, while 1.0 will never prune.) */
+            void setPruneThreshold(const double pp)
```
Updated the pruning threshold to a calculation that does not require any searching of the tree to calculate. A trade off is that this threshold will always prune on your first solution as the change in cost is relative to an infinite cost...

gammell commented on src/ompl/geometric/planners/rrt/RRTstar.h:
```diff
@@ -148,42 +160,125 @@ namespace ompl
+            }
+
+            /** \brief Controls whether heuristic rejection is used on samples (e.g., x_rand) */
+            void setSampleRejection(const bool reject);
```
Added the ability to do sample rejection. Previously this was **also** done if pruning was on.
I would argue that there is no need for this to exist in parallel with informed sampling, but I didn't want to be so presumptuous as to propose that. :smile:

isucan replied:
Can you expand on this a bit? Do you mean pruning should not be needed if informed sampling is available?

gammell replied:
Sorry, that statement was a bit vague. I meant there's no (long-term) need to support both rejection sampling and informed sampling of an admissible heuristic, as

1. Informed sampling is better than rejection sampling (working on a paper to show this), and
2. The way we wrote the classes, an objective without informed sampling just defaults to rejection sampling.

For now, I do think this should remain an option though in case someone wants to reject on a inadmissible heuristic.

gammell commented on src/ompl/geometric/planners/rrt/RRTstar.h:
```diff
@@ -148,42 +160,125 @@ namespace ompl
+            }
+
+            /** \brief Controls where heuristic rejection is used on new states before connection (e.g., x_new = steer(x_nearest, x_rand)) */
+            void setNewStateRejection(const bool reject)
```
Added the ability to specifically control new-state rejection. Previously this was done automatically if pruning was on.

gammell commented on src/ompl/geometric/planners/rrt/RRTstar.h:
```diff
@@ -148,42 +160,125 @@ namespace ompl
+            In RRT* this is accomplished by turning on Informed RRT* (tree pruning and focused sampling) and new-state rejection.
+            This flag individually sets the options described above.
+            */
+            void setFocusSearch(const bool focus)
```
This is the "meta" parameter to enable the generalized concept of focused search to facilitate use with CFOREST. Apparently I was less humble here and decided that the sample focusing would come from the informed sampling instead of rejection sampling.

gammell commented on src/ompl/geometric/planners/rrt/RRTstar.h:
```diff
@@ -148,42 +160,126 @@ namespace ompl
+                Considering the descendents of a vertex prevents removing a descendent
+                that may actually be capable of later providing a better solution once
+                its incoming path passes through a different vertex (e.g., a change in homotopy class). */
+            void setTreePruning(const bool prune)
```
So I realised that half the point of this to be able to propose changes and easily compare them. I've rewritten the pruning function to be less aggressive. Open to discussion if that's what you want.

gammell commented on src/ompl/geometric/planners/rrt/RRTstar.h:
```diff
@@ -148,42 +160,126 @@ namespace ompl
+            bool getInformedRrtStar() const
+            {
+                return useInformedSampling_ && getTreePruning();
+            }
```
Informed RRT\* is (technically) the shrinking of the problem domain during search, which requires a way to generate samples in the subdomain (informed sampling) and pruning the tree. As the tree pruning is now less aggressive, I can just use that in parallel and call the combination Informed RRT\*

isucan replied:
I think it is easier for users if they can construct a planner type Foo, rather than RRTstar and then set a bunch of things to make it actually be InformedRRTStar. How about we add an InformedRRTStar, which is like PRMstar -- essentially just a new name + one or two settings in the base class (so all the code remains in RRTstar)

mamoll replied:
I like Ioan's idea. It'll be easier for people to find InformedRRT* in the documentation.

gammell replied:
This is up to you.
I personally just think of Informed RRT\* as a demonstration of the informed sampling (and more generally the closed-form for path-length's informed subset), but I can easily make a meta-class version.

gammell commented on src/ompl/geometric/planners/rrt/RRTstar.h:
```diff
@@ -251,19 +356,22 @@ namespace ompl
-            /** \brief Computes the Cost To Go heuristically as the cost to come from start to motion plus
-                 the cost to go from motion to goal. If \e shortest is true, the estimated cost to come
-                 start-motion is given. Otherwise, this cost to come is the current motion cost. */
-            base::Cost costToGo(const Motion *motion, const bool shortest = true) const;
```
This is now unnecessary, see in the .cpp file.

gammell commented on src/ompl/geometric/planners/rrt/src/RRTstar.cpp:
```diff
@@ -101,6 +129,16 @@ void ompl::geometric::RRTstar::setup()
+            {
+                OMPL_INFORM("%s: No cost-to-go heuristic set. Defaulting to the distance from the state to the goal.", getName().c_str());
+                //Bind to the goalRegionCostToGo function.
+                opt_->setCostToGoHeuristic( boost::bind(&base::goalRegionCostToGo, _1, _2) );
```
I looked into `base::goalRegionCostToGo` and now fully understand what it's doing. It was (obviously) the correct thing to be using, but it makes even more sense to only use it if a heuristic has not otherwise been specified in `opt_`.

isucan replied:
Should we not move this code as part of setup() for the OptimizationObjective? Or at least for the ProblemDefinition?

mamoll replied:
I like Ioan's idea.

gammell replied:
I'm not sure I fully understood this comment.
`goalRegionCostToGo` assumes that the cost-to-go of a state is equal to the distance the state is from the goal region, which means it's only applicable to `PathLengthOptimizationObjective`.

I've moved this call to that constructor, is that what was desired?

gammell commented on src/ompl/geometric/planners/rrt/src/RRTstar.cpp:
```diff
@@ -207,9 +261,8 @@ ompl::base::PlannerStatus ompl::geometric::RRTstar::solve(const base::PlannerTer
-
-            if (prune_ && opt_->isCostBetterThan(bestCost_, costToGo(rmotion)))
+            // Attempt to generate a sample, if we fail (e.g., too many rejection attempts), skip the remainder of this loop and return to try again
+            if (!sampleUniform(rstate))
```
The rejection sampling that was being performed by the line `opt_->isCostBetterThan(bestCost_, costToGo(rmotion))` is now being done by the default rejection sampler that was recently introduced.

`sampleUniform` looks after calling the appropriate sampler and returns false if the sampler times out.

gammell commented on src/ompl/geometric/planners/rrt/src/RRTstar.cpp:
```diff
@@ -338,16 +394,15 @@ ompl::base::PlannerStatus ompl::geometric::RRTstar::solve(const base::PlannerTer
+            if (useNewStateRejection_)
             {
-                if (opt_->isCostBetterThan(costToGo(motion, false), bestCost_))
+                if (opt_->isCostBetterThan(solutionHeuristic(motion, false), bestCost_))
```
In the literature, this isn't really the same thing as pruning. This is a rejection scheme that is performed on the output of the steer function. Made it a separate setting.

gammell commented on src/ompl/geometric/planners/rrt/src/RRTstar.cpp:
```diff
@@ -338,16 +394,15 @@ ompl::base::PlannerStatus ompl::geometric::RRTstar::solve(const base::PlannerTer
                 }
                 else // If the new motion does not improve the best cost it is ignored.
                 {
-                    --statesGenerated;
```
You can't ungenerate a state. :wink:
It wasn't used, but all the work of generating it was done... Feel free to disagree.

isucan replied:
please check where this is used. Does this not report how many states were additionally stored in the tree? I suspect we should keep the --

gammell replied:
This variable is only used in the `OMPL_INFORM` line at the end of solve as `[...] Created %u new states [...]`. 

I guess what is the interpretation of the phrase creating a new state?

gammell commented on src/ompl/geometric/planners/rrt/src/RRTstar.cpp:
```diff
@@ -573,73 +661,210 @@ void ompl::geometric::RRTstar::getPlannerData(base::PlannerData &data) const
 }
 
-ompl::base::Cost ompl::geometric::RRTstar::costToGo(const Motion *motion, const bool shortest) const
+void ompl::geometric::RRTstar::allocSampler()
```
Create the requested sampler, either informed (which can be direct), rejection, or regular.
Since we're doing rejection like this anyway, I think we could do just informed or regular, but as I said, I wasn't ready to go that far.

isucan replied:
I am inclined to agree. Can you see any disadvantages to this?

gammell replied:
I don't think so?...

gammell commented on src/ompl/geometric/planners/rrt/src/RRTstar.cpp:
```diff
@@ -573,73 +661,210 @@ void ompl::geometric::RRTstar::getPlannerData(base::PlannerData &data) const
 }
 
-int ompl::geometric::RRTstar::pruneTree(const base::Cost pruneTreeCost)
+int ompl::geometric::RRTstar::pruneTree(const base::Cost& pruneTreeCost)
```
This is a more 'admissible' pruning scheme.

What it does is iterate through the graph and flag every "leaf vertex" that meets the pruning condition (i.e., cannot provide a better solution). It then deletes up the branches from these leaves for as long as the vertices are (a) cannot provide a better solution and (b) a leaf.

gammell commented on src/ompl/geometric/planners/rrt/src/RRTstar.cpp:
```diff
@@ -573,73 +661,210 @@ void ompl::geometric::RRTstar::getPlannerData(base::PlannerData &data) const
+        std::list<Motion*> leavesToPrune;
+
+        // Get the list from the NN structure
+        nn_->list(treeVertices);
```
I didn't want to start from startMotion_, since if we had multiple start states, we could have grown multiple independent trees.

gammell commented on src/ompl/geometric/planners/rrt/src/RRTstar.cpp:
```diff
@@ -573,73 +661,210 @@ void ompl::geometric::RRTstar::getPlannerData(base::PlannerData &data) const
+                leavesToPrune.push_back(*vIter);
+            }
+            // No else, does not meet the pruning criteria
+        }
```
This finds all the initial leaves to trim.

gammell commented on src/ompl/geometric/planners/rrt/src/RRTstar.cpp:
```diff
@@ -573,73 +661,210 @@ void ompl::geometric::RRTstar::getPlannerData(base::PlannerData &data) const
+
+            // Update our counter
+            ++numPruned;
+        }
```
This iterates through the list of leaves to trim, working from the front, we

1. Remove the leaf from the NN struct
2. Disconnecting it from its parent
3. Check if the parent should now be added to the list, (i.e., is it now a leaf? is it's heuristic estimate too high?) and adding if necessary.
4. Freeing the state and pointer
5. Removing from the list and getting the next iterator.

gammell commented on src/ompl/geometric/planners/rrt/src/RRTstar.cpp:
```diff
@@ -573,73 +661,210 @@ void ompl::geometric::RRTstar::getPlannerData(base::PlannerData &data) const
-        nn_->add(pruneScratchSpace_.newTree);
 
-        return (tree_size - pruneScratchSpace_.newTree.size());
+bool ompl::geometric::RRTstar::leafPruningCondition(Motion *motion, const base::Cost& pruneTreeCost)
```
The pruning condition in one place for ease of use on both the vertex and the vertex's parent.

gammell commented on src/ompl/geometric/planners/rrt/src/RRTstar.cpp:
```diff
@@ -573,73 +661,210 @@ void ompl::geometric::RRTstar::getPlannerData(base::PlannerData &data) const
 }
 
-void ompl::geometric::RRTstar::deleteBranch(Motion *motion)
+void ompl::geometric::RRTstar::setSampleRejection(const bool reject)
```
If we set sample rejection after a sampler has been allocated, we need to create a new sampler.
If a sampler has not yet been allocated, we don't need to.

gammell commented on src/ompl/geometric/planners/rrt/src/RRTstar.cpp:
```diff
@@ -573,73 +661,210 @@ void ompl::geometric::RRTstar::getPlannerData(base::PlannerData &data) const
+    }
+}
+
+void ompl::geometric::RRTstar::setInformedRrtStar(bool informedRrtStar)
```
Same as setRejectionSampler, allocate if necessary, defer otherwise.

gammell commented on src/ompl/geometric/planners/rrt/src/RRTstar.cpp:
```diff
@@ -573,73 +661,210 @@ void ompl::geometric::RRTstar::getPlannerData(base::PlannerData &data) const
+    }
+}
+
+bool ompl::geometric::RRTstar::sampleUniform(base::State *statePtr)
```
Use the appropriate sampler and return the result.

gammell commented:
So in taking a closer look at the pruning method and the base::goalRegionCostToGo, I took it upon myself to clean somethings up about `RRT*`. I hope I didn't step on anyone's toes in doing so, if I did I apologize.

Obviously I'm expecting some back and forth on whether they're the right choices. :wink:

isucan commented:
My usual request for rebasing :) I know this is a pain, but it is much easier to look at one commit rather than 40 to see what was done (when looking through history)

isucan commented on src/ompl/geometric/planners/bitstar/BITstar.h:
```diff
@@ -397,9 +407,6 @@ namespace ompl
             ///////////////////////////////////////
             // Planner progress property functions
             /** \brief Retrieve the best exact-solution cost found
-            as the raw data. (bestCost_) */
```
dropSample() above: please pass const& instead of by value

gammell replied:
I'm going to defer.
I can tell you a fun story sometime about a day spent learning the hard way that it is safer for delete functions for a shared pointer to be called by value. :wink:

Essentially, a delete function that does nontrivial cleanup (i.e., both accessing the pointer and removing it from owning containers) to delete the pointer before it's done with it.

I'm not sure if it's this exact function that requires the pointer by value, but one of the `BIT*` functions definitely does, so I used the same pattern for all of them to be safe.

mamoll commented on src/ompl/base/src/Planner.cpp:
```diff
@@ -134,6 +134,11 @@ ompl::base::PlannerStatus ompl::base::Planner::solve(double solveTime)
+ompl::base::PlannerStatus ompl::base::Planner::solve(time::duration solveTime)
+{
+    return solve(timedPlannerTerminationCondition(solveTime));
+}
```
I don't understand why `solve(double)` and `solve(time::duration)` are both needed and do slightly different things. If they can really have the same behavior, then you can pass in `solve(time::seconds(solveTime.seconds))` to the existing method.

gammell replied:
I dunno, I guess I just like syntactic-sugar... Undone.

isucan commented on src/ompl/geometric/planners/rrt/src/RRTstar.cpp:
```diff
@@ -512,20 +574,37 @@ ompl::base::PlannerStatus ompl::geometric::RRTstar::solve(const base::PlannerTer
+    }
+    else
+    {
+        double r = std::min(maxDistance_, r_rrg_*std::pow(log((double)(nn_->size() + 1u))/((double)(nn_->size() + 1u)), 1/(double)(si_->getStateDimension())));
```
white space around operators? * /

mamoll commented on src/ompl/base/src/PlannerTerminationCondition.cpp:
```diff
@@ -222,7 +222,12 @@ ompl::base::PlannerTerminationCondition ompl::base::plannerAndTerminationConditi
+
+ompl::base::PlannerTerminationCondition ompl::base::timedPlannerTerminationCondition(time::duration duration)
+{
+    return PlannerTerminationCondition(boost::bind(&timePassed, time::now() + duration));
```
This is also a case where we have two methods that really do the same thing. Can we just pick one way and stick with that?

gammell replied:
Well here's a case where I thought they fit nicely on top of each other.
The existing `PTC(double)` call converts the value into a `duration` internally, so all I was doing was breaking out that intermediate step. Then if one was using `duration`, they didn't have to convert to a `double` just to have the `PTC` convert back to a `duration`... That seemed likely to lose timing resolution.

isucan commented on src/ompl/geometric/planners/rrt/src/RRTstar.cpp:
```diff
@@ -573,73 +652,322 @@ void ompl::geometric::RRTstar::getPlannerData(base::PlannerData &data) const
+    // r_rrg > 2*(1+1/d)^(1/d)*(measure/ballvolume)^(1/d)
+    if (useInformedSampling_ == true)
+    {
+        r_rrg_ = rewireFactor_*2.0*std::pow((1.0 + 1.0/dimDbl)*(prunedInfMeasure_/unitNBallMeasure(si_->getStateDimension())), 1.0/dimDbl);
```
white space around operators?

gammell replied:
In that you want them, correct?

isucan commented:
Pending minor issues & rebase, this is ready for merge in my opinion! Thank you for your contribution!

mamoll commented on src/ompl/geometric/planners/rrt/src/RRTstar.cpp:
```diff
@@ -230,6 +277,7 @@ ompl::base::PlannerStatus ompl::geometric::RRTstar::solve(const base::PlannerTer
         }
 
         // Check if the motion between the nearest state and the state to add is valid
+        ++collisionChecks_;
```
Is this really the number of calls to the state validity checker? If not, does each collision check take the same amount of time? I'm curious what the intended use is of this variable.

gammell replied:
It is the number of calls to `SpaceInformation::checkMotion`.
The intent was as a planner progress property, as it can be interesting to see how many collision checks a planner is making. I thought it was useful, if your opinion differs, that's fine just let me know. :smile:

mamoll replied:
`SpaceInformation::checkMotion` can correspond to several collision checks. If you want the number of collision checks, you'd need to count the number of calls to the various StateValidityChecker::isValid(...) methods. It's easy enough to keep track of this with an internal counter in the StateValidityChecker, but I'm not sure we should make this a requirement for a StateValidityChecker-derived class. But suppose you do this, then you can update ompl::base::Planner::getPlannerData() to add the number of collision checks as an extra property.

gammell replied:
So's that a remove it since the information isn't meaningful?

mamoll replied:
I don't think it is meaningful. Worse, people might think it literally means number of collision checks, which it doesn't. In other words, I think you should remove it.

mamoll commented on src/ompl/geometric/planners/rrt/src/RRTstar.cpp:
```diff
@@ -573,73 +652,322 @@ void ompl::geometric::RRTstar::getPlannerData(base::PlannerData &data) const
+        // The list of leaves to prune
+        std::list<Motion*> leavesToPrune;
+        // The list of chain vertices to recheck after pruning
+        std::list<Motion*> chainsToRecheck;
```
Did you benchmark this pruning scheme against the current scheme? In general, std::list is slow as you are allocating elements one at a time. In the current scheme we use std::vector::reserve to do just one alloc. Also, better memory coherence with std::vector. I think the Scott Meyer books on effective C++ have some interesting info on this.

gammell replied:
No, I didn't benchmark this against the previous scheme as the previous pruning method was inadmissible. By removing a vertex independent of the heuristic-value of its children, the pruning method does not maintain uniform density in the subproblem, and more importantly can remove vertices that could be part of a better solution (e.g., in the attached picture, the highlighted vertex and its descendents would be removed by the old method, even though some of the descendents will be part of a better solution once a passage is found through the gap).

![output](https://cloud.githubusercontent.com/assets/5250910/9202206/c009a366-401e-11e5-8fb5-e5c2aaff9301.png)


Regardless, you are correct that in the previous method that `std::vector` was allocated in a single call, but the difference is that it did so twice, both of the size of the NN structure (it made `std::vector`s for the vertices to keep and prune, respectively).

Alternatively, this method does not store the vertices to keep in a `std::vector`, but instead immediately adds them back to the NN structure. So no allocation for those vertices ever occurs. The other method didn't do this because it was actually counting the vertices to decide whether to go through with the prune or not (see other comment).

The only vertices it does store are ones that it cannot immediately make a decision on (`chainsToRecheck`) and those that are to be pruned (`leavesToPrune`). For these vertices, it's just using the `std::list` as a queue...

I guess the underlying assumption is that you will be pruning (significantly) less vertices than the size of the graph. Would preallocating to the entire size of the tree still be beneficial? I don't really know.

mamoll replied:
Ok, let's leave things as you have them for now. Try running ompl_benchmark on some simple problems with the old RRT* and new RRT* code (both with pruning on) to make sure nothing broke and, ideally, that the new version is faster.

As a general rule of thumb, it's almost always better to use std::vector over std::list *unless* you frequently need to remove elements. Instead of mentioning "memory coherence" I should said that std::vector's are more cache friendly. It may not matter much if you still have to deference Motion*'s, so I am willing to leave it the way you have it, provided it doesn't make RRT* significantly slower.

isucan replied:
-1 for std::list. It is very rarely the case that a list is faster, even with remove operations. I am also ok to leave this is in if benchmarking shows it makes no difference.

mamoll commented on src/ompl/geometric/planners/rrt/src/RRTstar.cpp:
```diff
@@ -573,73 +652,322 @@ void ompl::geometric::RRTstar::getPlannerData(base::PlannerData &data) const
+        std::list<Motion*> chainsToRecheck;
+
+        //Clear the NN structure:
+        nn_->clear();
```
It could be expensive to rebuild the entire nn structure each time you call pruneTree. Each insertion is probably O(d*log(n)), where d is the number of dimensions of the c-space and n the number of elements in the nn structure. Of course this depends on the nn structure (it could be even worse) and the distance function.

gammell replied:
I kept the pattern of rebuilding the NN structure from scratch as that was was the existing prune method. The alternative would be to call `remove` on each vertex to be removed...

mamoll commented on src/ompl/geometric/planners/rrt/src/RRTstar.cpp:
```diff
@@ -573,73 +652,322 @@ void ompl::geometric::RRTstar::getPlannerData(base::PlannerData &data) const
 
-    // To create the new nn takes one order of magnitude in time more than just checking how many
-    // states would be pruned. Therefore, only prune if it removes a significant amount of states.
-    if ((double)pruneScratchSpace_.newTree.size() / tree_size < pruneStatesThreshold_)
```
I seemed to remember that this was also important. @jvgomez?

gammell replied:
@jvgomez, please correct me if I'm wrong, but my understanding is this is a condition to decide whether it is worth the effort to prune or not. It is calculated from the number of vertices that will be pruned, and as a result requires evaluating the entire tree to make a prune decision on each vertex.

The proposed change replaces this condition with one that is based on the change in cost since the last time the graph was pruned.

The advantage to a cost-based method is that it doesn't require any calculation (just a comparison of `Cost`) before deciding to prune. The disadvantage is that it doesn't know how many vertices will actually be pruned from the tree.

gammell commented:
Just about to push changes as per comments.

Is it okay if I delay rebasing until the PR is ready to be merged? I believe there are still some outstanding items to discuss.

gammell commented on src/ompl/geometric/planners/rrt/src/RRTstar.cpp:
```diff
@@ -573,73 +645,400 @@ void ompl::geometric::RRTstar::getPlannerData(base::PlannerData &data) const
 
-    // To create the new nn takes one order of magnitude in time more than just checking how many
-    // states would be pruned. Therefore, only prune if it removes a significant amount of states.
-    if ((double)pruneScratchSpace_.newTree.size() / tree_size < pruneStatesThreshold_)
```
GitHub has changed the diff it associates for this line, unfortunately burying the previous conversation about it. This is a comment to draw attention to that.

gammell commented:
Ok,
Resurrecting this PR after my absence.
I've changed the `std::list` structures that were of concern to `std::queue`-wrapped `std::deque`s. From what I [understand](http://stackoverflow.com/questions/1262808/which-stl-container-should-i-use-for-a-fifo), the underlying `deque` has the cache performance that you desired, and for me it keeps the intent of the code obvious.

I wrote some code that ran normal RRT* to get a good, dense graph and then called the prune algorithm. Testing `std::queue`s with both `std::list` and `std::deque` gave average prune times of `1.0568e-3`sec and `7.965e-4`sec, respectively (on an old computer). So it appears that your intuition for `std::list` was correct.

Let me know if this addresses your concerns or if I misunderstood them. I'll get to work on fixing that unit test issue now.

mamoll commented:
I have checked the last couple commits and they look good to me. It's hard to navigate through all the comments on different versions of this PR, but it seems like we are now ready to merge, right?

gammell commented:
Yeah, is there anything I can do to make it easier to navigate through all the different versions? I've rebased it.

There's one PR I'd like to submit after this is merged, it updates Informed Sampling and BIT* to handle move-it and multiple starts/goals.

mamoll commented:
Once you resolve the conflicts, I'll merge this one.

gammell commented:
Huh, those are new...
I was just about to start packing up for the day, but I will sort this out first. Give me a second.

gammell commented:
I think I got it, but give me a second to the merge a once-over.

gammell commented:
Yup, I'm pretty sure that's it. 
The conflict was just the fix I made to the intermediateSolutionCallback in the first PR, I should have rebased this branch onto that one yesterday in preparation, sorry.

I'll get my last PR up this weekend, but right now I've got to run.
